keras/utils/vis_utils.py

```diff
@@ -88,6 +88,11 @@ def model_to_dot(model,
         dot.set('dpi', dpi)
         dot.set_node_defaults(shape='record')
 
+    sub_n_first_node = {}
```

Comment:
This is a nice improvement over the previous version. But the main problem is that the submodels with multiple input and/or outputs are incorrectly represented when their first and last nodes are retained. 

Proper way would be to keep track of the inputs and outputs of each wrapper/submodule in these dictionaries.

Reply:
Thanks for pointing out the problem. I partly corrected it by allowing submodels to submodels links. However, submodels with multiple inputs and outputs needs more work and I am not sure whether I am going to do it since I personally don't need them right now.